Production DB support (postgres)

finn-block opened the pull request:
This PR defines a `Storage` interface and adds support for choosing between Bolt and Postgres storage types.

* BoltDB code has been mostly left as is, with modifications of adding unused context arguments to function signatures to make it conform to the interface, making functions that don't conform to the interface private (to discourage skipping the interface) and removing functions that are unused.
* Postgres support uses [sqlc](https://docs.sqlc.dev/en/latest/) to generate query wrappers and [goose](https://pkg.go.dev/github.com/pressly/goose/v3) for migrations.
* A switch statement takes a URI for the storage provider, and returns either bolt or postgres based on the URI scheme. This creates a backwards-incompatible change requiring the configured Bolt path to be prefixed with `bolt://`. If that is an issue, adding a special case mapping of no scheme -> bolt would be pretty easy. I also renamed the config field from `db_file` to `storage_uri`.

Fixes #41

finn-block committed:
Add initial postgres migration and a few queries

not called from anywhere yet

finn-block committed:
move stuff around to generally work

postgres wants the db opened for each request and closed after
bolt wants it opened once at server startup and closed on server shutdown

this will take some work

finn-block committed:
implement postgres functions

finn-block committed:
make the tests pass

finn-block committed:
minor pg fixes

finn-block committed:
add golangci-lint

finn-block committed:
fix pg schema

finn-block committed:
rename DBFile config field name

thereisnogabe commented on impl/config/config.go:
```diff
@@ -71,7 +71,7 @@ func GetDefaultConfig() Config {
 			LogLocation: "log",
 			LogLevel:    "debug",
-			DBFile:      "diddht.db",
+			DBFile:      "bolt://diddht.db",
```
this might be problematic since unix won't be able to create files of this format

finn-block replied:
see PR description (updated at almost the exact time you left this message): this config option is parsed as a URI and the scheme is used to determine what storage backend to use. The default behavior is unchanged, it still uses bolt with a file called `diddht.db` in the current directory.

thereisnogabe replied:
ok cool!

thereisnogabe commented on impl/pkg/storage/pkarr/pkarr.go:
```diff
@@ -0,0 +1,17 @@
+	Seq int64  `json:"seq" validate:"required"`
+}
+
+type PKARRStorage interface {
```
I did this but we should rename this to `PkarrStorage` for consistency, since `Pkarr` is how the project is supposed to be represented.

finn-block replied:
hrm, I was actually planning to rename it to `Storage` because it's already in a package called `pkarr`. Is there a different term we could use? If not, I'm happy to just name it `Pkarr` and reference it as `pkarr.Pkarr`

thereisnogabe replied:
naming it storage makes sense to me

finn-block replied:
Went to rename it and i realized this isnt actually used anywhere, so i just removed it. I did rename the only other thing in this file (`PKARRRecord` -> `Record`)

thereisnogabe commented on impl/pkg/storage/db/bolt/pkarr.go:
```diff
@@ -1,32 +1,19 @@
+	"github.com/TBD54566975/did-dht-method/pkg/storage/pkarr"
 )
 
 const (
```
should this also move to `impl/pkg/storage/pkarr/pkarr.go`?

alternatively could move it (and the contents of the new `pkarr.go`) to a new files `impl/pkg/storage/models.go`

finn-block replied:
this file is the bolt specific implementation of `storage.Storage`. I think it used to be in a file called `pkarr.go` so i just kept the name. I will rearrange it a bit be more clear what it's doing.

thereisnogabe commented on impl/pkg/storage/db/postgres/db.go:
```diff
@@ -0,0 +1,32 @@
+)
+
+type DBTX interface {
+	Exec(context.Context, string, ...interface{}) (pgconn.CommandTag, error)
+	Query(context.Context, string, ...interface{}) (pgx.Rows, error)
+	QueryRow(context.Context, string, ...interface{}) pgx.Row
```
```suggestion
	Exec(context.Context, string, ...any) (pgconn.CommandTag, error)
	Query(context.Context, string, ...any) (pgx.Rows, error)
	QueryRow(context.Context, string, ...any) pgx.Row
```

thereisnogabe replied:
ah I see this is generated, disregard

thereisnogabe commented on impl/pkg/storage/db/postgres/postgres.go:
```diff
@@ -0,0 +1,120 @@
+
+type postgres string
+
+func NewPostgres(uri string) (postgres, error) {
```
public method returns private type!

finn-block replied:
It implements a public interface (`storage.Storage`) that we can't import into this package without creating circular imports. Would you prefer `postgres` type be made public? I definitely see how this can be confusing, but IMO just documenting that the return type implements `storage.Storage` should be sufficient.

thereisnogabe replied:
yes maybe add a note as such, all good

thereisnogabe commented on impl/pkg/storage/pkarr_test.go:
```diff
@@ -1,19 +1,31 @@
+	if err != nil {
+		panic(err)
+	}
+	// defer db.Close()
```
uncomment?

thereisnogabe commented on impl/pkg/storage/pkarr_test.go:
```diff
@@ -1,19 +1,31 @@
+
+	db, err := storage.NewStorage(uri)
+	if err != nil {
+		panic(err)
```
since it's a test you can do `t.RequireNoError()`

thereisnogabe commented on impl/pkg/storage/storage.go:
```diff
@@ -1,180 +1,36 @@
+	case "postgres":
+		return postgres.NewPostgres(uri)
+	default:
+		return nil, fmt.Errorf("unsupported db type %s (from uri %s)", u.Scheme, uri)
```
let's make bolt the default case too so we don't fail

finn-block replied:
How about just adding bolt for the case where the scheme is `""` (which usually means a file path was provided, not a URI)? It would be pretty confusing for ie. `mysql://user:pass@localhost/diddht` to produce a boltdb file rather than an error.

thereisnogabe replied:
sgtm

thereisnogabe reviewed (approved):
awesome work. some small non-blocking comments.

I also think it would be nice to have updated docs on how this works. Can anyone just run postgres locally? any setup needed?

finn-block committed:
Consolidate bolt down to one file to make it more like postgres implementation

Also drop a bunch of unused bolt functions and address a few other misc feedback

finn-block committed:
update NewX documentation on storage implementations
